src/Http/Authentication/AuthenticateRequest.php

```
@@ -41,12 +49,42 @@ public function process(ServerRequestInterface $request, RequestHandlerInterface
             return $handler->handle($request);
         }
 
+        $clientId = $this->getClientId($request);
+
+        if ($clientId !== null) {
+            return $this->handleClientId($request, $handler, $clientId);
+        }
+
         $apiKey = $this->getApiKey($request);
 
         if (empty($apiKey)) {
             return (new MissingCredentials())->toResponse();
         }
 
+        return $this->handleApiKey($request, $handler, $apiKey);
+    }
+
+    private function handleClientId(ServerRequestInterface $request, RequestHandlerInterface $handler, string $clientId): ResponseInterface
+    {
+        try {
+            $metadata = $this->auth0Client->getMetadata($clientId, $this->auth0Client->getToken());
+        } catch (Exception $exception) {
```

I remember that we said not to reject a client id when Auth0 is down. Is there a way to differentiate exceptions here between "the client does not exist" and "auth0 is down"?

---

I would assume that if auth0 is down we don't get a 4xx error, but either a 5xx or no real response at all (not sure how Guzzle handles that)

---

I will try to simulate this with an invalid endpoint or url.

---

Since it can be hard to predict what we will get back when Auth0 is actually down (vs an invalid endpoint), we could maybe only consider the client id invalid if the response code is in the 4xx range? Since that indicates specifically that there is a problem with the client id, instead of a problem with Auth0 itself.

---

I tried with an invalid domain and `ConnectionException` seems to be the desired exception to implement this bypass of the validation

---

Is that also for responses that have a 5xx status?
Because I think we need to account for either Auth0 not being reachable at all (connection exceptions), and Auth0 returning internal errors (500, 503, ...)

---

Oh and maybe timeout exceptions are also another kind?

---

We could also take another approach and add a configuration to disable apiKey and/or clientId validation?

---

That already exists I think? But it requires manual intervention then (both to disable and re-enable). I think that we should be good if we disable the automatic exception throwing from Guzzle, and then check the response code ourselves. If it's 404 we return a Forbidden (client does not exist), if it's 200 we check the metadata like we do now, and if it's anything else we temporarily allow it? And we should then still have a try/catch around it for connection exceptions I think, because a timeout for example doesn't return a response I imagine.

---

I proposed the manual intervention to keep the code simple and avoid possible edge-cases were we would always allow access. So in case Auth0 goes down we have a "emergency plan" and disable the `Auth0` check inside `udb3-search`?

---

@bertramakers I reworked the code to closely match your proposal:
- I disabled Guzzle exceptions
- Everything else then a 200 response code returned from Auth0 is treated as missing metadata and results in an `InvalidClientId`
- A `ConnectionException` is still handled and is equivalent to Auth0 being down
- Other possible exceptions are ignored and bubble up to make sure coding/logical errors are not missed

---

@bertramakers as discussed the 500 HTTP errors are handled as ConnectionException